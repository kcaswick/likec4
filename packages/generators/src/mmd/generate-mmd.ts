import { Colors, DefaultThemeColor } from '@likec4/core'
import type { ComputedEdge, ComputedNode, ComputedView, NodeId } from '@likec4/core/types'
import { CompositeGeneratorNode, NL, joinToNode, toString } from 'langium'
import { isNil } from 'rambdax'

const capitalizeFirstLetter = (value: string) => value.charAt(0).toLocaleUpperCase() + value.slice(1)

const fqnName = (nodeId: string): string => nodeId.split('.').map(capitalizeFirstLetter).join('')

const nodeName = (node: ComputedNode): string => {
  return fqnName(node.parent ? node.id.slice(node.parent.length + 1) : node.id)
}

const mmdshape = ({ shape }: ComputedNode): [start: string, end: string] => {
  switch (shape) {
    case 'queue':
    case 'cylinder':
      return ['[(', ')]']
    case 'person': {
      return ['[fa:fa-user ', ']']
    }
    case 'storage':
      return ['([', '])']
    case 'mobile':
    case 'browser':
    case 'rectangle': {
      return ['[', ']']
    }
  }
}

export function generateMermaid<V extends ComputedView>(view: V) {
  const { nodes, edges } = view
  const names = new Map<NodeId, string>()

  const printNode = (node: ComputedNode, parentName?: string): CompositeGeneratorNode => {
    const name = nodeName(node)
    const fqnName = (parentName ? parentName + '.' : '') + name
    names.set(node.id, fqnName)

    const label = nodeLabel(node)
    const shape = mmdshape(node)

    const baseNode = new CompositeGeneratorNode()

    if (node.children.length > 0) {
      baseNode
        .append('subgraph ', fqnName, '[')
        .indent({
          indentedChildren: indent =>
            indent.append(label, ']', NL).appendIf(
              node.children.length > 0,
              NL,
              joinToNode(
                nodes.filter(n => n.parent === node.id),
                n => printNode(n, fqnName)
              )
            ),
          indentImmediately: false,
          indentation: 2
        })
        .append('end', NL)
    } else {
      baseNode.append(fqnName, shape[0], label, shape[1], NL)
    }

    baseNode.appendIf(
      node.color !== DefaultThemeColor,
      'style ',
      fqnName,
      ' fill:',
      Colors[node.color].fill,
      ',stroke:',
      Colors[node.color].stroke,
      ',color:',
      Colors[node.color].hiContrast,
      NL
    )
    return baseNode
  }
  //     return `${names.get(edge.source)} -> ${names.get(edge.target)}${edge.label ? ': ' + edge.label : ''}`
  const printEdge = (edge: ComputedEdge): CompositeGeneratorNode => {
    return new CompositeGeneratorNode().append(
      names.get(edge.source),
      ' --',
      edge.label ? '"' + edge.label.replaceAll('\n', '\\n') + '"--' : '',
      '> ',
      names.get(edge.target)
    )
  }

  return toString(
    new CompositeGeneratorNode()
      .appendIf(view.title !== null && view.title.length > 0, '---', NL, `title: ${view.title}`, NL, '---', NL)
      .append('graph ', view.autoLayout, NL, NL)
      .append(
        joinToNode(
          nodes.filter(n => isNil(n.parent)),
          n => printNode(n),
          {
            appendNewLineIfNotEmpty: true
          }
        )
      )
      .appendIf(
        edges.length > 0,
        NL,
        joinToNode(edges, e => printEdge(e), {
          appendNewLineIfNotEmpty: true
        })
      )
  )
}

function nodeLabel(node: ComputedNode) {
  // The following regexs are based on the Mermaid parser at https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/diagrams/flowchart/parser/flow.jison
  const RE_UNICODE_TEXT = /[\u0100-\uFFFF]/u // /[\p{L}--[\u0000-\u00A9]]/v;
  const RE_START_TEXT = /(?<=\(\[|\[\[|[^\-]>|\[\)|\(\(\(|\||\(|\[|\{)/u // Note that most mmdshapes are here, except ellipse and trapezoid which forbid more punctuation than the others
  const RE_CONTINUE_TEXT = /[^\[\]\(\)\{\}\|\"]+/ // This can only be used to continue text once started
  // Note: double quote was removed from the character class below from NODE_STRING - in the jison file other rules usually prevent it from matching
  // const RE_NODE_STRING = /(?:[A-Za-z0-9!\#$%&'*+\.`?\\_\/]|-(?=[^\>\-\.])|=(?!=))+/v;
  // const RE_textNoTags = new RegExp(
  //   `^(?:${RE_NODE_STRING.source}|${/* RE_UNICODE_TEXT.source */""}|${(/[\d]+|[\s&:*#]|\^|[xo<]?(?:--|==|-\.)/u).source})+$`,
  //   'u'
  // )
  const RE_textNoTags =
    /^(?:(?:[A-Za-z0-9!#$%&'*+\.`?\\_\/]|-(?=[^>\-\.])|=(?!=))+|[\d]+|[\s&:*#]|\^|[xo<]?(?:--|==|-\.))+$/u
  const RE_STR = /^"[^"]+"$/u
  const RE_MD_STR = /^\"\`[^`"]+\`\"$/

  const isMultiline = Boolean(node.technology || node.description)
  const isPlainText = RE_textNoTags.test(node.title)
  const hasUnicodeText = RE_UNICODE_TEXT.test(node.title)

  if (!isMultiline) {
    if (isPlainText && !hasUnicodeText) {
      return node.title.replaceAll('\n', '\\n')
    }
    if (RE_STR.test(`"${node.title}"`)) {
      return `"${node.title}"`
    }
  }

  const useColor = node.color !== DefaultThemeColor

  return new CompositeGeneratorNode()
    .append(`"`)
    .appendIf(useColor, `<span style='color: ${Colors[node.color].hiContrast}'>`)
    .indent({
      indentedChildren: indent =>
        indent
          .append(joinToNode(node.title.replace(`"`, `&quot;`).split('\n'), i => i, { separator: NL }))
          .appendIf(useColor, '</span>')
          .appendIf(
            Boolean(node.technology),
            n =>
              n
                .appendNewLineIfNotEmpty()
                .append(`<span style='font-size:0.6em`)
                .appendIf(useColor, `; color: ${Colors[node.color].loContrast}`).appendTemplate`'>${joinToNode(
                node.technology?.replace(`"`, `&quot;`)?.split('\n') ?? [],
                i => i,
                { separator: NL }
              )}</span>`
          )
          .appendIf(
            Boolean(node.description),
            n =>
              n
                .appendNewLineIfNotEmpty()
                .append(`<span style='font-size:0.7em`)
                .appendIf(useColor, `; color: ${Colors[node.color].loContrast}`).appendTemplate`'>${joinToNode(
                node.description?.replace(`"`, `&quot;`)?.split('\n') ?? [],
                i => i,
                { separator: NL }
              )}</span>`
          )
          .append(`"`),
      indentImmediately: false,
      indentation: 2
    })
}
