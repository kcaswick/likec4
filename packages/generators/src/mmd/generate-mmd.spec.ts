import { expect, test } from 'vitest'
import { fakeDiagram, fakeDiagram2, fakeComputedView3Levels, fakeWithColorsAndShapes } from '../__mocks__/data'
import { generateMermaid } from './generate-mmd'

test('generate mermaid - fakeDiagram', () => {
  expect(generateMermaid(fakeDiagram)).toMatchSnapshot()
})

test('generate mermaid - fakeDiagram2', () => {
  expect(generateMermaid(fakeDiagram2)).toMatchSnapshot()
})

test('generate mermaid - fakeComputedView 3 Levels', () => {
  expect(generateMermaid(fakeComputedView3Levels)).toMatchSnapshot()
})

test('generate mermaid - with colors and shapes', () => {
  expect(generateMermaid(fakeWithColorsAndShapes)).toMatchSnapshot()
})
