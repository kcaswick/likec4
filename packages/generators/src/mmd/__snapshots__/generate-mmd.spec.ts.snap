// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate mermaid - fakeComputedView 3 Levels 1`] = `
"graph 

Amazon[amazon]
subgraph Cloud[cloud]

  subgraph Cloud.Backend[backend]

    Cloud.Backend.Graphql[graphql]
    Cloud.Backend.Storage[storage]
  end
  Cloud.FrontendAdminPanel[adminPanel]
  Cloud.FrontendDashboard[dashboard]
end
Customer[customer]
Support[support]

Cloud.FrontendDashboard --> Cloud.Backend.Graphql
Cloud.FrontendAdminPanel --> Cloud.Backend.Graphql
Cloud.Backend.Storage --> Amazon
Cloud.Backend.Graphql --> Cloud.Backend.Storage
Support --> Cloud.FrontendAdminPanel
Customer --> Cloud.FrontendDashboard
"
`;

exports[`generate mermaid - fakeDiagram 1`] = `
"graph 

Amazon[amazon]
subgraph Cloud[cloud]

  Cloud.Backend[backend]
  Cloud.Frontend[frontend]
end
Customer[customer]
Support[support]

Cloud.Frontend --"requests"--> Cloud.Backend
Cloud.Backend --> Amazon
Support --> Cloud.Frontend
Customer --"opens"--> Cloud.Frontend
"
`;

exports[`generate mermaid - fakeDiagram2 1`] = `
"---
title: frontend
---
graph 

Client[client]
SystemBackend[backend]
SystemFrontend[frontend]

SystemFrontend --"requests"--> SystemBackend
Client --"opens"--> SystemFrontend
"
`;

exports[`generate mermaid - with colors and shapes 1`] = `
"---
title: Landscape view
---
graph TB

System[system]
Spa[SPA]
Customer[fa:fa-user Customer]
style Customer fill:#6366f1,stroke:#4f46e5,color:#eef2ff
Mobile[Mobile]
style Mobile fill:#16a34a,stroke:#15803d,color:#111827
subgraph Lambdas["Lambdas
<span style='font-size:0.7em'>Serverless compute</span>"]

  Lambdas.Fn_enrich["Enrichment
<span style='font-size:0.7em'>Enriches raw data
by adding additional information</span>"]
end
Pg([PostgreSQL])
subgraph Sqs["SQS
<span style='font-size:0.7em'>Simple Queue Service</span>"]

  Sqs.Queue1[("Raw Data
<span style='font-size:0.7em'>Queue with raw data</span>")]
  Sqs.Queue2[("<span style='color: #eef2ff'>Enriched Data</span>
<span style='font-size:0.7em; color: #c7d2fe'>Filtered and preprocessed data
Ready for further processing</span>")]
  style Sqs.Queue2 fill:#6366f1,stroke:#4f46e5,color:#eef2ff
end

Lambdas.Fn_enrich --"writes enriched data"--> Sqs.Queue2
Lambdas.Fn_enrich --"reads raw data"--> Sqs.Queue1
Lambdas.Fn_enrich --"reads additional information"--> Pg
Customer --> Mobile
"
`;
